crates/ty_project/src/db.rs

```diff
@@ -126,6 +137,64 @@ impl ProjectDatabase {
         )
     }
 
+    /// Deserialize the database from the persistent cache.
+    pub fn deserialize(&mut self, path: &str, metadata: &ProjectMetadata) -> anyhow::Result<()> {
+        // Read from the persistent cache.
+        let contents = match std::fs::read(path) {
+            Ok(contents) => contents,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(()),
+            Err(err) => return Err(err.into()),
+        };
+
+        let _value_source = ValueSourceGuard::new(ValueSource::Cli, false);
+        let _lint_registry = LintRegistryGuard::new(|name| {
+            default_lints_registry()
+                .get(name)
+                .map(|lint| lint.name())
+                .ok()
+        });
+
+        // Deserialize the database.
+        let mut decoder = bincode::serde::BorrowedSerdeDecoder::from_slice(
+            &contents,
+            bincode::config::standard(),
+            (),
+        );
+
+        <dyn salsa::Database>::deserialize(self, decoder.as_deserializer())?;
+
+        // Sync any deserialized inputs.
+        for file in File::load_all(self) {
+            file.sync(self);
+            self.files.seed(file, self);
+        }
+
+        for root in FileRoot::load_all(self) {
+            self.files.seed_root(root, self);
+        }
+
+        for project in Project::load_all(self) {
+            if project.resolve_settings(metadata, self)? {
```

Comment:
Nit: `resolve_settings` seems to be only used by deserialization but its name doesn't suggest that. Maybe `init_after_deserialization`?